### PR TITLE
v: fix aliased type checking

### LINKED ISSUE #18278: No type error when use type alias for `int`
### Describe the bug

Code: https://vosca.dev/p/535b0da2cf

```v
type Int = int

struct Foo {
	value Int
}

fn take_u8(u u8) {
	println(u)
}

fn main() {
	foo := Foo{
		value: 10000
	}
	take_u8(foo.value)
}
```

### Expected Behavior

Error: cannot use Int as u8

### Current Behavior

Output:

```
16
```

### Reproduction Steps

Run code above

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 fc4c431.e738d67

### Environment details (OS name and version, etc.)

```
V full version: V 0.3.4 fc4c431.6db62e4
OS: linux, Ubuntu 22.04.2 LTS
Processor: 2 cpus, 64bit, little endian, Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz

getwd: /home/pmakhnev/playground
vexe: /home/pmakhnev/v/v
vexe mtime: 2023-05-26 06:00:08

vroot: OK, value: /home/pmakhnev/v
VMODULES: OK, value: /root/.vmodules
VTMP: OK, value: /tmp/v_0

Git version: git version 2.34.1
Git vroot status: weekly.2023.21-24-g6db62e43
.git/config present: true

CC version: cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

### PR DESCRIPTION
**Pull request by felipensp**
Fix #18278

Because this change the compiler has spotted too many places where int was used when t.Type(u32) was expected.

Several places where ast.Type was used as `int` it now requires proper casting `ast.idx_to_type(idx)`.

Before this PR `int` was allowed where aliased `u32` type is expected.

#### Added
- `table.find_type()`
- `Type.idx_type()`

image
image
image

### CODE REVIEW

File: vlib/v/checker/assign.v [hunk 1]
```diff
@@ -717,10 +717,10 @@ or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.',
 				modified_left_type := if !left_type.is_int() {
 					c.error('invalid operation: shift on type `${c.table.sym(left_type).name}`',
 						node.pos)
-					ast.void_type_idx
+					ast.idx_to_type(ast.void_type_idx)
```
Comment (spytheman) [inline vlib/v/checker/assign.v:720]:
```suggestion
					ast.void_type
```

File: vlib/v/checker/assign.v [hunk 2]
```diff
@@ -717,10 +717,10 @@ or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.',
...
 				} else if left_type.is_int_literal() {
 					// int literal => i64
-					ast.u32_type_idx
+					ast.idx_to_type(ast.u32_type_idx)
```
Comment (spytheman) [inline vlib/v/checker/assign.v:723]:
```suggestion
					ast.u32_type
```

File: vlib/v/checker/infix.v [hunk 3]
```diff
@@ -676,10 +676,10 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 			modified_left_type := if !left_type.is_int() {
 				c.error('invalid operation: shift on type `${c.table.sym(left_type).name}`',
 					left_pos)
-				ast.void_type_idx
+				ast.idx_to_type(ast.void_type_idx)
```
Comment (spytheman) [inline vlib/v/checker/infix.v:679]:
```suggestion
				ast.void_type
```

File: vlib/v/checker/infix.v [hunk 4]
```diff
@@ -676,10 +676,10 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
...
 			} else if left_type.is_int_literal() {
 				// int literal => i64
-				ast.u32_type_idx
+				ast.idx_to_type(ast.u32_type_idx)
```
Comment (spytheman) [inline vlib/v/checker/infix.v:682]:
```suggestion
				ast.u32_type
```

File: vlib/v/gen/c/auto_str_methods.v [hunk 5]
```diff
@@ -1244,7 +1244,7 @@ fn should_use_indent_func(kind ast.Kind) bool {
 	return kind in [.struct, .alias, .array, .array_fixed, .map, .sum_type, .interface]
 }
 
-fn (mut g Gen) get_enum_type_idx_from_fn_name(fn_name string) (string, int) {
+fn (mut g Gen) get_enum_type_idx_from_fn_name(fn_name string) (string, ast.Type) {
```
Comment (spytheman) [inline vlib/v/gen/c/auto_str_methods.v:1247]: the name is `get_enum_type_idx`, it should return int; if you make it return ast.Type, change the name too

File: vlib/v/gen/c/auto_str_methods.v [hunk 6]
```diff
@@ -1265,8 +1265,7 @@ fn (mut g Gen) get_enum_type_idx_from_fn_name(fn_name string) (string, int) {
 	return mod_enum_name, idx
 }
 
-fn (mut g Gen) gen_enum_static_from_string(fn_name string, mod_enum_name string, idx int) {
-	enum_typ := ast.idx_to_type(idx)
+fn (mut g Gen) gen_enum_static_from_string(fn_name string, mod_enum_name string, enum_typ ast.Type) {
```
Comment (spytheman) [inline vlib/v/gen/c/auto_str_methods.v:1268]: the call to that, do pass `idx`:
```
vlib/v/gen/c/fn.v:2192:g.gen_enum_static_from_string(fn_name, mod_enum_name, idx)
```

File: vlib/v/gen/c/cgen.v [hunk 7]
```diff
@@ -4057,13 +4057,13 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
...
-					g.write('${g.typ(node.expr_type.idx())}_${m.name}')
+					g.write('${g.typ(ast.idx_to_type(node.expr_type.idx()))}_${m.name}')
 				}
 				return
 			}
 			receiver := m.params[0]
-			expr_styp := g.typ(node.expr_type.idx())
-			data_styp := g.typ(receiver.typ.idx())
+			expr_styp := g.typ(ast.idx_to_type(node.expr_type.idx()))
+			data_styp := g.typ(ast.idx_to_type(receiver.typ.idx()))
```
Comment (spytheman) [inline vlib/v/gen/c/cgen.v:4066]: given how often there is `ast.idx_to_type(t.idx())`, perhaps we can make also add
`t.idx_type()`, that does it directly, to simplify all the callers

Comment (spytheman) [reply vlib/v/gen/c/cgen.v:4066]: i.e. `pub fn (t Type) idx_type() Type {`

File: vlib/v/gen/c/infix.v [hunk 8]
```diff
@@ -720,7 +720,7 @@ fn (mut g Gen) infix_expr_is_op(node ast.InfixExpr) {
 				ast.idx_to_type(g.table.type_idxs['None__'])
 			}
 			else {
-				ast.no_type
+				ast.idx_to_type(ast.no_type)
```
Comment (spytheman) [inline vlib/v/gen/c/infix.v:723]: that was already an ast.Type

Comment (spytheman) [reply vlib/v/gen/c/infix.v:723]: why call ast.idx_to_type() to it again?

File: vlib/v/gen/js/js.v [hunk 9]
```diff
@@ -3104,7 +3104,7 @@ fn (mut g JsGen) gen_infix_expr(it ast.InfixExpr) {
 			g.gen_deref_ptr(it.right_type)
 			g.write(')')
 		} else {
-			mut greater_typ := 0
+			mut greater_typ := ast.Type(0)
```
Comment (spytheman) [inline vlib/v/gen/js/js.v:3107]: `ast.no_type`

Comment (spytheman) [inline vlib/v/gen/js/js.v:3107]:
```suggestion
			mut greater_typ := ast.no_type
```

File: vlib/v/gen/native/expr.v [hunk 10]
```diff
@@ -366,7 +366,7 @@ fn (mut g Gen) gen_print_from_expr(expr ast.Expr, typ ast.Type, name string) {
 			}
 		}
 		ast.OffsetOf {
-			styp := g.typ(expr.struct_type)
+			styp := g.typ(ast.idx_to_type(expr.struct_type))
```
Comment (spytheman) [inline vlib/v/gen/native/expr.v:369]: why is that needed? `expr.struct_type` for ast.OffsetOf is already ast.Type

File: vlib/v/gen/native/gen.v [hunk 11]
```diff
@@ -826,7 +826,7 @@ fn (mut g Gen) get_sizeof_ident(ident ast.Ident) i32 {
 		ast.Var { ident.obj.typ }
 	}
 	if typ != 0 {
-		return g.get_type_size(typ)
+		return g.get_type_size(ast.idx_to_type(typ))
```
Comment (spytheman) [inline vlib/v/gen/native/gen.v:829]: change the name to `typ_idx`

Comment (spytheman) [reply vlib/v/gen/native/gen.v:829]: or just `idx`

**Comment by spytheman** [comment]
Please also add a `_test.v` file, that checks the problem from the issue.

File: vlib/v/gen/c/infix.v (hunk 8 above)
Comment (felipensp) [reply vlib/v/gen/c/infix.v:723]: oooops :-)

**Review by spytheman — APPROVED** [review]
Excellent work.
